src/triggers.js

```diff
@@ -393,44 +408,39 @@ export function maybeRunQueryTrigger(triggerType, className, restWhere, restOpti
 // Resolves to an object, empty or containing an object key. A beforeSave
 // trigger will set the object key to the rest format object to save.
 // originalParseObject is optional, we only need that for before/afterSave functions
-export function maybeRunTrigger(triggerType, auth, parseObject, originalParseObject, config) {
+export function maybeRunTrigger(triggerType, auth, parseObject, originalParseObject, config, context) {
   if (!parseObject) {
     return Promise.resolve({});
   }
   return new Promise(function (resolve, reject) {
     var trigger = getTrigger(parseObject.className, triggerType, config.applicationId);
     if (!trigger) return resolve();
-    var request = getRequestObject(triggerType, auth, parseObject, originalParseObject, config);
-    var response = getResponseObject(request, (object) => {
+    var request = getRequestObject(triggerType, auth, parseObject, originalParseObject, config, context);
+    var { success, error } = getResponseObject(request, (object) => {
       logTriggerSuccessBeforeHook(
         triggerType, parseObject.className, parseObject.toJSON(), object, auth);
+      if (triggerType === Types.beforeSave || triggerType === Types.afterSave) {
+        Object.assign(context, request.context);
+      }
       resolve(object);
     }, (error) => {
       logTriggerErrorBeforeHook(
         triggerType, parseObject.className, parseObject.toJSON(), auth, error);
       reject(error);
     });
-    // Force the current Parse app before the trigger
-    Parse.applicationId = config.applicationId;
-    Parse.javascriptKey = config.javascriptKey || '';
-    Parse.masterKey = config.masterKey;
 
     // AfterSave and afterDelete triggers can return a promise, which if they
     // do, needs to be resolved before this promise is resolved,
     // so trigger execution is synced with RestWrite.execute() call.
     // If triggers do not return a promise, they can run async code parallel
     // to the RestWrite.execute() call.
-    var triggerPromise = trigger(request, response);
-    if(triggerType === Types.afterSave || triggerType === Types.afterDelete)
-    {
-      logTriggerAfterHook(triggerType, parseObject.className, parseObject.toJSON(), auth);
-      if(triggerPromise && typeof triggerPromise.then === "function") {
-        return triggerPromise.then(resolve, resolve);
+    return Promise.resolve().then(() => {
```

Comment:
is this just to make async?  if so, could use Process.nextTick() instead???  or am i missing something again.

Reply:
it's to wrap to everything that is thrown inside is gonna be rejected as errorrs support if  ` promise`  isn't a `Promise` but a directly returned value AND is promise is a `Promise`.

This avoids all the checks + using `try  {} catch (e) {}` 

Reply:
got it.